relay/nakama/siwe/siwe.go

```diff
@@ -0,0 +1,276 @@
+package siwe
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"math"
+	"math/big"
+	"slices"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/rotisserie/eris"
+	"github.com/spruceid/siwe-go"
+
+	"pkg.world.dev/world-engine/relay/nakama/utils"
+)
+
+const (
+	// TODO: This static domain and URI should be configurable via an environment variable.
+	DefaultDomain = "example.com"
+	DefaultURI    = "https://example.com/v2/account/authenticate/custom"
+
+	// DefaultTimeout is the length of time between the issue time and the expiration time. If the current time
+	// is not between the issue time and expiration time of a message, the signature will be rejected.
+	DefaultTimeout   = 5 * time.Minute
+	DefaultStatement = "Log in to Nakama using SIWE."
+
+	siweMessageCollection = "siwe_message_collection"
+)
+
+var (
+	ErrMissingSignerAddress = errors.New("missing signer address")
+	ErrMissingMessage       = errors.New("message field is required")
+	ErrMissingSignature     = errors.New("signature field is required")
+
+	maxNonce = big.NewInt(0).SetUint64(math.MaxUint64)
+)
+
+// GenerateResult will be serialized to JSON and sent back to a client. The client must sign the message in the
+// SIWEMessage field, and resubmit their authentication request.
+type GenerateResult struct {
+	SIWEMessage string `json:"siwe_message"`
+}
+
+// nonceStorageObj is the struct that is serizlied to JSON and saved to the Nakama storage layer. Each signer address
+// maps to a different nonceStorageObj. A signature is valid if the nonce contained in the signed message does NOT
+// already have an entry in this object.
+type nonceStorageObj struct {
```

Comment:
on a separate note, we should probably consolidate all Nakama storage stuff into a single `storage` package

Reply:
i would also note here that this is not to be confused with the cardinal tx nonce

Reply:
I added a triage ticket to consolidate the common storage object operations: https://linear.app/arguslabs/issue/WORLD-991/consolidate-nakama-storage-object-operations.

I'll add a note about the cardinal tx nonces.